src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/FileIO/FileSystem.vb

```diff
@@ -0,0 +1,2034 @@
+
+            Debug.Assert(System.Enum.IsDefined(GetType(CopyOrMove), operation), "Invalid Operation!!!")
+            Debug.Assert(sourceDirectoryPath <> "" And IO.Path.IsPathRooted(sourceDirectoryPath), "Invalid Source!!!")
+            Debug.Assert(targetDirectoryPath <> "" And IO.Path.IsPathRooted(targetDirectoryPath), "Invalid Target!!!")
+
+            ' Special case for moving: If target directory does not exist, AND both directories are on same drive,
+            '   use IO.Directory.Move for performance gain (not copying).
+            If operation = CopyOrMove.Move And Not IO.Directory.Exists(targetDirectoryPath) And
+                IsOnSameDrive(sourceDirectoryPath, targetDirectoryPath) Then
+
+                ' Create the target's parent. IO.Directory.CreateDirectory won't throw if it exists.
+                IO.Directory.CreateDirectory(GetParentPath(targetDirectoryPath))
+
+                Try
+                    IO.Directory.Move(sourceDirectoryPath, targetDirectoryPath)
+                    Exit Sub
+                Catch ex As IO.IOException
+                Catch ex As UnauthorizedAccessException
+                    ' Ignore IO.Directory.Move specific exceptions here. Try to do as much as possible later.
+                End Try
+            End If
+
+            ' Create the target, create the root node, and call the recursive function.
+            System.IO.Directory.CreateDirectory(targetDirectoryPath)
+            Debug.Assert(IO.Directory.Exists(targetDirectoryPath), "Should be able to create Target Directory!!!")
+
+            Dim SourceDirectoryNode As New DirectoryNode(sourceDirectoryPath, targetDirectoryPath)
+            Dim Exceptions As New ListDictionary
+            CopyOrMoveDirectoryNode(operation, SourceDirectoryNode, overwrite, Exceptions)
+
+            ' Throw the final exception if there were exceptions during copy / move.
+            If Exceptions.Count > 0 Then
+                Dim IOException As New IO.IOException(Utils.GetResourceString(SR.IO_CopyMoveRecursive))
+                For Each Entry As DictionaryEntry In Exceptions
+                    IOException.Data.Add(Entry.Key, Entry.Value)
+                Next
+                Throw IOException
+            End If
+        End Sub
+
+        ''' <summary>
+        ''' Given a directory node, copy or move that directory tree.
+        ''' </summary>
+        ''' <param name="Operation">Specify whether to move or copy the directories.</param>
+        ''' <param name="SourceDirectoryNode">The source node. Only copy / move directories contained in the source node.</param>
+        ''' <param name="Overwrite">True to overwrite sub-files. Otherwise False.</param>
+        ''' <param name="Exceptions">The list of accumulated exceptions while doing the copy / move</param>
+        <ResourceExposure(ResourceScope.Machin
```

Comment:
In the reference source, we're calling `NativeMethods.MoveFileEx()` here. What is the difference between `MoveFileEx()` and `Delete()` + `Move()`?

Reply:
I am not sure where you are looking, the lines above have nothing to do with MoveFileEx. Delete/Move works everywhere, CopyOrMoveFile line 1138 does did us MoveFileEx but only under certain conditions,  MoveFileEx only works on Win32NT, I will put the if back but not sure what to do with CodeAccessPermission.RevertAssert()